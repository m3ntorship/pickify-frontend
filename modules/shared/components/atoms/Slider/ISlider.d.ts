declare namespace ISlider {
  export interface IProps {
    type: 'circular' | 'horizontal' | 'vertical';
    progress: number;
    verticalMeterHeight?: number;
    meterColor?: 'error' | 'primary-shd5' | 'primary';
    radius?: number;
  }
}

export { ISlider };
