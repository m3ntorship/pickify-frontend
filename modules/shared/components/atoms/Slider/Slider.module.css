.outerHorizontal {
  @apply w-full h-3 bg-white flex items-center p-1 rounded-sm;
}

.innerHorizontal {
  @apply h-1 rounded-sm;
}

.success {
  @apply bg-success;
}

.primary {
  @apply bg-primary;
}

.primary-shd5 {
  @apply bg-primary-shd5;
}

.error {
  @apply bg-error;
}

.innerVertical {
  @apply w-2 bg-white;
}

.outerVertical {
  @apply w-2 rounded-sm;
}

.wrapper {
  @apply rounded-sm py-1 px-2 bg-white w-3 flex justify-center;
}

.circular {
  @apply -rotate-90 origin-center stroke-current;
}

.primary-text {
  @apply text-primary;
}

.error-text {
  @apply text-error;
}
