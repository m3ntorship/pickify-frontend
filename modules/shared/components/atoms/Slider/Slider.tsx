import React from 'react';
import classNames from 'classnames';
import type { FC, ReactElement } from 'react';
import type { ISlider } from './ISlider';
import styles from './Slider.module.css';

const imageHeight = 100;

const Slider: FC<ISlider.IProps> = ({
  progress,
  type = 'horizontal',
  verticalMeterHeight = imageHeight,
  meterColor = 'primary',
  radius,
}): ReactElement => {
  const maximumProgress = 100;
  const height = (progress * verticalMeterHeight) / maximumProgress;
  const innerHorizontalClasses = classNames([styles['inner-horizontal']], {
    [styles.success]: progress === maximumProgress,
    [styles.primary]: progress < maximumProgress,
  });
  const innerVerticalClasses = classNames([styles['inner-vertical']], {
    [styles.primary]: meterColor === 'primary',
    [styles['primary-shd5']]: meterColor === 'primary-shd5',
    [styles.error]: meterColor === 'error',
  });
  const circularClasses = classNames([styles.circular], {
    [styles['primary-text']]: meterColor === 'primary',
    [styles['error-text']]: meterColor === 'error',
  });
  const circular = {
    radius: 0,
    diameter: 0,
    stroke: 7,
    normalizedRadius: 0,
    normalizedDiameter: 0,
    circumference: 0,
    doubleStroke: 0,
  };
  const defaultRadius = 60;
  circular.radius = radius ?? defaultRadius;
  circular.diameter = circular.radius + circular.radius;
  circular.doubleStroke = circular.stroke + circular.stroke;
  circular.normalizedRadius = circular.radius - circular.doubleStroke;
  circular.normalizedDiameter =
    circular.normalizedRadius + circular.normalizedRadius;
  circular.circumference = circular.normalizedDiameter * Math.PI;

  if (type === 'horizontal') {
    return (
      <div className={`${styles['outer-horizontal']} ${styles.wrapper}`}>
        <div
          className={innerHorizontalClasses}
          style={{ width: `${progress}%` }}
        />
      </div>
    );
  }
  if (type === 'vertical') {
    return (
      <div className="flex h-full">
        <div
          className={`${styles['outer-vertical']} ${styles.wrapper}`}
          style={{ height: `${height}px` }}
        >
          <div className={innerVerticalClasses} />
        </div>
      </div>
    );
  }
  return (
    <svg width={circular.diameter} height={circular.diameter}>
      <circle
        className={circularClasses}
        strokeWidth={circular.stroke}
        strokeDasharray={`${circular.circumference} ${circular.circumference}`}
        style={{
          strokeDashoffset: `${
            circular.circumference -
            (progress / maximumProgress) * circular.circumference
          }`,
        }}
        fill="transparent"
        r={circular.normalizedRadius}
        cx={circular.radius}
        cy={circular.radius}
      />
    </svg>
  );
};

export default Slider;
