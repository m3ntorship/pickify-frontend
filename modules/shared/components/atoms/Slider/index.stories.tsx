import type { ReactElement } from 'react';
import type { Story } from '@storybook/react';
import Slider from '.';
import type { ISlider } from './ISlider';

export default {
  title: 'atoms/slider',
  component: Slider,
  argTypes: {
    type: {
      control: {
        type: 'radio',
        options: ['horizontal', 'vertical', 'circular'],
      },
    },
    progress: {
      control: {
        type: 'range',
        min: 0,
        max: 100,
        step: 1,
      },
    },
    height: {
      control: {
        type: 'number',
      },
    },
    radius: {
      control: {
        type: 'number',
      },
    },
    verticalMeterColor: {
      control: {
        type: 'radio',
        options: ['primary', 'primary-shd5', 'error'],
      },
    },
  },
};

const Template: Story<ISlider.IProps> = (args): ReactElement => (
  <Slider {...args} />
);

export const sliderStory = Template.bind({});
sliderStory.args = {
  progress: 90,
  type: 'vertical',
  height: 300,
  radius: 60,
  verticalMeterColor: 'primary',
};
