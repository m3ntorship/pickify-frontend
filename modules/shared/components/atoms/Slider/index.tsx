import type { FC, ReactElement } from 'react';
import React from 'react';
import classNames from 'classnames';
import type { ISlider } from './ISlider';

const Slider: FC<ISlider.IProps> = ({
  progress,
  type,
  height,
  verticalMeterColor,
  radius,
}): ReactElement => {
  const maximumProgress = 100;
  const halfMaximumProgress = 50;
  const defaultHeight = 300;
  const outerHorizontalClasses = classNames(
    'w-full h-3 bg-white flex items-center p-1 rounded-sm',
  );
  const innerHorizontalClasses = classNames('h-1 rounded-sm', {
    'bg-success': progress === maximumProgress,
    'bg-primary': progress < maximumProgress,
  });
  const outerVerticalClasses = classNames('w-2 rounded-sm', {
    'bg-success': progress === maximumProgress,
    'bg-primary':
      progress < maximumProgress && verticalMeterColor === 'primary',
    'bg-primary-shd5':
      progress < maximumProgress && verticalMeterColor === 'primary-shd5',
    'bg-error': progress < maximumProgress && verticalMeterColor === 'error',
  });
  const innerVerticalClasses = classNames('w-2 bg-white rounded-t-sm');
  const wrapperClasses = classNames(
    'rounded-sm py-1 px-2 bg-white w-3 flex justify-center',
  );
  const circularClasses = classNames(
    '-rotate-90 origin-center stroke-current',
    {
      'text-primary': progress > halfMaximumProgress,
      'text-error': progress < halfMaximumProgress,
    },
  );
  const circular = {
    radius: 0,
    diameter: 0,
    stroke: 4,
    normalizedRadius: 0,
    circumference: 0,
    doubleStroke: 0,
  };
  const defaultRadius = 60;
  circular.radius = radius ?? defaultRadius;
  circular.diameter = circular.radius + circular.radius;
  circular.doubleStroke = circular.stroke + circular.stroke;
  circular.normalizedRadius = circular.radius - circular.doubleStroke;
  circular.circumference = circular.diameter * Math.PI;
  if (type === 'horizontal') {
    return (
      <div className={outerHorizontalClasses}>
        <div
          className={innerHorizontalClasses}
          style={{ width: `${progress}%` }}
        />
      </div>
    );
  }
  if (type === 'vertical') {
    return (
      <div className={wrapperClasses}>
        <div
          style={{ height: `${height ?? defaultHeight}px` }}
          className={outerVerticalClasses}
        >
          <div
            className={innerVerticalClasses}
            style={{ height: `${maximumProgress - progress}%` }}
          />
        </div>
      </div>
    );
  }
  return (
    <svg width={circular.diameter} height={circular.diameter}>
      <circle
        className={circularClasses}
        strokeWidth={circular.stroke}
        strokeDasharray={`${circular.circumference} ${circular.circumference}`}
        style={{
          strokeDashoffset: `${
            circular.circumference -
            (progress / maximumProgress) * circular.circumference
          }`,
        }}
        fill="transparent"
        r={circular.radius - circular.doubleStroke}
        cx={circular.radius}
        cy={circular.radius}
      />
    </svg>
  );
};

export default Slider;
