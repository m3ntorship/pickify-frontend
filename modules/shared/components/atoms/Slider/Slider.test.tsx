import React from 'react';
import renderer from 'react-test-renderer';
import Slider from './Slider';

describe('snapshot testing', () => {
  it('should render horizontally when given horizontal property', () => {
    const tree = renderer
      .create(<Slider type="horizontal" progress={50} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render vertically when given vertical property', () => {
    const tree = renderer
      .create(
        <Slider
          type="vertical"
          progress={50}
          verticalMeterHeight={300}
          meterColor="primary"
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render circular when given circular property', () => {
    const tree = renderer
      .create(<Slider type="circular" progress={50} radius={50} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
