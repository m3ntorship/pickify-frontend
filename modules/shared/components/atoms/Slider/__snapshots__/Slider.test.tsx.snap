// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing should render circular when given circular property 1`] = `
<svg
  height={100}
  width={100}
>
  <circle
    className="circular primary-text"
    cx={50}
    cy={50}
    fill="transparent"
    r={36}
    strokeDasharray="226.1946710584651 226.1946710584651"
    strokeWidth={7}
    style={
      Object {
        "strokeDashoffset": "113.09733552923255",
      }
    }
  />
</svg>
`;

exports[`snapshot testing should render horizontally when given horizontal property 1`] = `
<div
  className="outer-horizontal wrapper"
>
  <div
    className="inner-horizontal primary"
    style={
      Object {
        "width": "50%",
      }
    }
  />
</div>
`;

exports[`snapshot testing should render vertically when given vertical property 1`] = `
<div
  className="flex h-full"
>
  <div
    className="outer-vertical wrapper"
    style={
      Object {
        "height": "150%",
      }
    }
  >
    <div
      className="inner-vertical primary"
    />
  </div>
</div>
`;
